### ISSUE TITLE
vktrace of Dota2 on Linux does not work.

### ISSUE THREAD

**Issue by davidlunarg** [LunarG/VulkanTools#121]
The command:
`vktrace -p dota2 -o d.vktrace -a '-vulkan -vktrace'`
results in:
`vktrace errror: Recieved signal SIGSEGV on non-mapped memory. Fatal error.`
Also, when running with debug build, get the error:
`Assertion got != g_deviceDatamap.end() failed.`
Tracing dota2 with PMB disabled seems to work:
 `vktrace -p dota2 -o d.vktrace -a '-vulkan -vktrace +vulkan_buffer_update_mode 0' --PMB false`

**Comment by cnorthrop** [comment]
I'm getting the same SIGSEGV when trying to trace Talos on Android.  I think we need to hook up code that continues to respect the application's existing signal handler.

**Comment by eero-t** [comment]
Note that vktrace signal handler is buggy, it uses malloc when it does output.  That can easily corrupt program memory.  "man 7 signal" lists which functions are safe to use in signal handler context.

**Comment by eero-t** [comment]
For me DOTA2 (today's update in Steam) crashes to same error even when I set set '-vulkan -vktrace +vulkan_buffer_update_mode 0' in Steam DOTA2 launch options, and run vktrace server as:
vktrace  -o dota2.vktrace --PMB false

(This is with today's Git version of Mesa Intel Vulkan driver on SKL GT2 machine with Ubuntu 16.04.)

Is that expected to work?

**Comment by davidlunarg** [comment]
eero-t: I fixed this crash yesterday. Make sure your source of VulkanTools includes commit a888a4bb5ffdcf0842f7aa148e70026a9c340d15.

If you still have problems with that commit in the source, please let me know. Note that trace of DOTA2 w/o +vulkan_buffer_update_mode 0 and --PMB false does not work. It's a know problem, and I'm working on it.

Also note that running DOTA2 with the -vktrace option has problems in the 12/13/16 version of DOTA2. Some characters are drawn completely black.

**Comment by eero-t** [comment]
I've updated DOTA2 to today's version and compiled today's GIT versions of VulkanTools (6fb9bf1cfce) and Mesa (47351b843a82).

Tracing (with server) worked now fine without crashes, but replay of the trace gives (when DOTA2 main "menu" comes up):
```
vkreplay error: Return value VK_ERROR_OUT_OF_DATE_KHR from API call (vkAcquireNextImageKHR) does not match return value from trace file VK_SUCCESS.
vkreplay error: Failed to replay packet_id 163, with global_packet_index 58751.
vkreplay error: Return value VK_ERROR_OUT_OF_DATE_KHR from API call (vkAcquireNextImageKHR) does not match return value from trace file VK_SUCCESS.
vkreplay error: Failed to replay packet_id 163, with global_packet_index 61121.
Segmentation fault (core dumped)
```

I've set all DOTA2 graphics options to max, except for shadows which is just set to high.  Is replay supposed to work with some other DOTA2 gfx options?

Note: Without "-vktrace" option given to DOTA2, rendering is seriously broken.  Without "+vulkan_buffer_update_mode 0" option, there's trace error when the game itself would start.  "--PBM" option doesn't seem to affect trace size at all, so setting it false is fine.  With those options, I still got once signal 6 abort.

**Comment by eero-t** [comment]
Updated to latest VulkanTools (e35a1aa1ba) and Mesa (ede8c02ab0).

Now I get signal 6 crash for Steam always when using the "--PBM false" option.  Without that, DOTA2 tracing works fine (resulting trace might also be smaller, but the size seems to vary a lot).

With the Mesa Intel Vulkan driver changes discussed in issue 124, I'm able to replay the resulting trace to the end, although replay outputs the 2 errors listed above.  Without those small Mesa changes, vkreplay crashes early in the trace like earlier.

While replay rendering looks fine for the DOTA2 main screen, the graphics in the actual (tutorial) game play look broken.  Validation layer shows a lot of these kind of errors on replay:
* invalid Semaphores
* vkCmdCopyBufferToImage(): pRegion[0] bufferOffset 0x92ee must be a multiple of 4

I don't see any of these errors when using validation for DOTA2 directly.

While tracing, either occlusion doesn't always work, or it messes up order in which frames are show (old content flickers on screen).  These issues happen only when tracing is enabled.

**Comment by davidlunarg** [comment]
I have several changes currently being reviewed that allow DOTA2 to be traced and replayed successfully. Hope to push them to master later today.

**Comment by davidlunarg** [comment]
Fixed in commit 2734a064f06a72032fde44242b76f15893fdd3ac.
Note that trace/replay of Dota2 on Windows with PMB enabled still does not work. Will submit a separate issue on that problem.

**Comment by eero-t** [comment]
What are the DOTA2 and tracing options with which one is supposed to get fully working traces?

Using latest DOTA2 and git tip of Mesa (Intel) Vulkan driver (4d6d4f9) & VulkanTools (2d2ce6149):
* DOTA2 traces are now 25-50% larger than week ago (regardless of --PMB option), so I've had trouble getting full ones
* If I use "--PMB true" and run the Monkey King demo, DOTA2 crashes with: "vktrace error: mprotect sys call failed. Fatal error."  At that point trace file size was >30GB, and it hadn't yet reach the actual demo game play
* When trace layer is enabled, there are some drawing errors (black rectangles or rubbish textures) and screen drawing frequently (wrongly) repeats earlier frames.  This happens both during tracing & replay. Validation still shows errors listed in my previous comment (invalid semaphores etc)
* Because tracing & replay show same rendering errors and tracing produces gigantic trace files, I tried using /dev/null as trace output to investigate the rendering errors more (without constantly running out of SSD space).  Once it crashed to "vktrace error: SIGSEGV received on the same address. Fatal error", once it worked.  There seem to be less errors with /dev/null output, so the rendering errors during tracing could be timing related (=broken semaphores?)

**Comment by eero-t** [comment]
Tried with today's version of Mesa, VulkanTools & DOTA2
* Don't see anymore garbage textures with tracing, only the frame update issue (related to DRI3?)
* Trace sizes are down to what they were two weeks ago
* Trace replay deadlocks every time right after showing couple of errors.  Those seem timing related as replay works fine if one runs it under "strace", and when run under Valgrind, replay dies to error

**Comment by eero-t** [comment]
Today's DOTA2, Mesa & VulkanTools:
* Tracing everything in DOTA2 menus works fine as does its replay (no deadlock freezes at replay startup anymore)
* However, when one starts tutorial / game, that *never finishes loading when tracing is enabled*, despite waiting until trace file was ~200GB
* If  (core) validation layer is enabled, replay crashes, whereas actual game is fine with validation.  These are last error before crash:
```
DS(ERROR): object: 0x1 type: 5 location: 4791 msgCode: 77: Queue 0x0xeeb9b8 is waiting on semaphore 0x1 that has no way to be signaled.
vkreplay error: Return value VK_ERROR_OUT_OF_DATE_KHR from API call (vkAcquireNextImageKHR) does not match return value from trace file VK_SUCCESS.
vkreplay error: Failed to replay packet_id 163, with global_packet_index 53112.
vkreplay error: Return value VK_ERROR_OUT_OF_DATE_KHR from API call (vkAcquireNextImageKHR) does not match return value from trace file VK_SUCCESS.
vkreplay error: Failed to replay packet_id 163, with global_packet_index 69239.
DS(ERROR): object: 0xf8fb40 type: 27 location: 12170 msgCode: 90: vkQueuePresentKHR: Swapchain image index too large (64). There are only 3 images in this swapchain.
Segmentation fault (core dumped)
```

I think this should be opened until there are clear instructions, how one should be able to successfully trace & replay DOTA2.

**Comment by stroyan** [comment]
Were you using both core validation and unique_objects layers when replaying?
That "waiting on semaphore 0x1" looks like the mesa driver returned multiple semaphore handles of 0x1 that were never mapped to unique handles.

The VK_ERROR_OUT_OF_DATE_KHR is going to be a deeper problem.  The driver is indicating that the swapchain has become invalid during replay in a way that did not happen during tracing.

**Comment by eero-t** [comment]
I get that semaphore error + crash with just "VK_LAYER_LUNARG_core_validation".

With full validation, I get:
```
DS(ERROR): object: 0xa4 type: 27 location: 12170 msgCode: 90: vkQueuePresentKHR: Swapchain image index too large (64). There are only 3 images in this swapchain.

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff56f9c77 in xshmfence_reset (f=0x19) at ../../src/xshmfence_futex.c:93
93	../../src/xshmfence_futex.c: No such file or directory.
(gdb) bt
#0  0x00007ffff56f9c77 in xshmfence_reset (f=0x19) at ../../src/xshmfence_futex.c:93
#1  0x00007ffff5f4b772 in x11_present_to_x11 (chain=0x9cfd20, image_index=<optimized out>, target_msc=0)
    at ../../../../src/vulkan/wsi/wsi_common_x11.c:702
#2  0x00007ffff5f4b459 in anv_QueuePresentKHR (_queue=0x9223f8, pPresentInfo=0x35d48d0) at ../../../../src/intel/vulkan/anv_wsi.c:375
#3  0x00007ffff44ddd4d in unique_objects::QueuePresentKHR(VkQueue_T*, VkPresentInfoKHR const*) ()
   from dbuild/layers/./libVkLayer_unique_objects.so
(gdb) up
#1  0x00007ffff5f4b772 in x11_present_to_x11 (chain=0x9cfd20, image_index=<optimized out>, target_msc=0)
    at ../../../../src/vulkan/wsi/wsi_common_x11.c:702
702	   xshmfence_reset(image->shm_fence);
686	static VkResult
687	x11_present_to_x11(struct x11_swapchain *chain, uint32_t image_index,
688	                   uint32_t target_msc)
689	{
690	   struct x11_image *image = &chain->images[image_index];
691	
692	   assert(image_index < chain->image_count);
693	
(gdb) 
694	   uint32_t options = XCB_PRESENT_OPTION_NONE;
695	
696	   int64_t divisor = 0;
697	   int64_t remainder = 0;
698	
699	   if (chain->base.present_mode == VK_PRESENT_MODE_IMMEDIATE_KHR)
700	      options |= XCB_PRESENT_OPTION_ASYNC;
701	
702	   xshmfence_reset(image->shm_fence);
(gdb) up
#2  0x00007ffff5f4b459 in anv_QueuePresentKHR (_queue=0x9223f8, pPresentInfo=0x35d48d0) at ../../../../src/intel/vulkan/anv_wsi.c:375
375	      result = swapchain->queue_present(swapchain,
376	                                        pPresentInfo->pImageIndices[i]);
(gdb) info locals
last = 0x7ffff44bf87b <__gthread_mutex_unlock(pthread_mutex_t*)+38>
i = 0
result = <optimized out>
(gdb) print pPresentInfo->pImageIndices[0]
$1 = 64
```
-> verifies that segfault is due to the last error.

With core validation, segfault doesn't happen every time.  When it does, only difference is that caller is "core_validation::QueuePresentKHR()" instead of "unique_objects::QueuePresentKHR()", and it happens from different layer SO.

Maybe difference is just timing, core validation is a bit faster than full one, so it's (in this case) less likely to hit the issue?   Which could also explain why it doesn't happen without validation, as validation seems to add quite a bit of overhead...


Btw. When I was running replay under Valgrind, I got following at end (besides lots of uninitialized memory  warnings from Mesa Vulkan driver which looked like just missing Valgrind allocation annotations):
```
==5265== Mismatched free() / delete / delete []
==5265==    at 0x4C2F24B: operator delete(void*) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==5265==    by 0x411083: vktrace_replay::Sequencer::~Sequencer() (vkreplay_seq.h:57)
==5265==    by 0x410D71: vkreplay_main(int, char**, unsigned int) (vkreplay_main.cpp:395)
==5265==    by 0x410E08: main (vkreplay_main.cpp:555)
==5265==  Address 0x1346e0b0 is 0 bytes inside a block of size 104 alloc'd
==5265==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==5265==    by 0x4172AC: vktrace_malloc (vktrace_memory.h:39)
==5265==    by 0x417E5B: vktrace_read_trace_packet (vktrace_trace_packet_utils.c:263)
==5265==    by 0x41177B: vktrace_replay::Sequencer::get_next_packet() (vkreplay_seq.cpp:34)
==5265==    by 0x4102BA: vktrace_replay::main_loop(vktrace_replay::ReplayDisplay, vktrace_replay::Sequencer&, vktrace_replay::vktrace_trace_packet_replay_library**, vkreplayer_settings) (vkreplay_main.cpp:91)
==5265==    by 0x410CAE: vkreplay_main(int, char**, unsigned int) (vkreplay_main.cpp:396)
```

**Comment by ardaLunarG** [comment]
Regarding to : VK_ERROR_OUT_OF_DATE_KHR only
This is happening in Intel Mesa driver when application window resizing. During resize new dimension swap chain images getting reallocated. If during this activity images were busy Mesa driver start to poll XCB events in the drivers wsi part and if this polled event dimensions does not match to the resize event dimensions sent by application VK_ERROR_OUT_OF_DATE_KHR is generated. Here driver tries to tell to the application to send resized dimensions again. Thus driver seems like doing the right thing.

There a few cases around this during Vkreplay:
**1- If user resize window during trace:** necessary swap chain API's getting collected and we do resize the window during replay, However in this case logs get flooded with vkreplay error: Return value VK_ERROR_OUT_OF_DATE_KHR from API call (vkAcquireNextImageKHR) does not match return value from trace file VK_SUCCESS. 
We need to debug why driver throw these during replay but not during trace.

**2- If user resize window during replay:** This causes a conflict with recorded swap chain APIs information in trace file, so we have to disable resizing during replay. (even user resize window we have to ignore and should not call swap chain allocation API's)

**3- Initial resize during replay:** We generate replay window 100x100 and then resize it to according to the traced app for the first time. This is regular behaviour of vkreplay. However in some cases even this initial resize fails ( happens very rare ) with same error: Return value VK_ERROR_OUT_OF_DATE_KHR from API call (vkAcquireNextImageKHR) does not match return value from trace file VK_SUCCESS. 

First and third case should be debugged considering driver or vkreplay might be doing something wrong. Second case disabling window resize should be the solution.

We are working on these now.

**Comment by ardaLunarG** [comment]
Regarding to:
==5265== Mismatched free() / delete / delete []

[branch](https://github.com/LunarG/VulkanTools/tree/vkreplayLeak) has the patch for the issue.
However as this patch still not preventing all the leaks in trace/replay I am still working on it.

**Comment by stroyan** [comment]
The SEGFAULT under vkQueuePresentKHR just looks like a consequence of carrying on after a VK_ERROR_OUT_OF_DATE_KHR result.
The message about "vkQueuePresentKHR: Swapchain image index too large (64)" could happen after a vkAcquireNextImageKHR errors out and vkReplay::replay calls m_objMapper.add_to_pImageIndex_map with a junk value for the image index.

**Comment by eero-t** [comment]
Today I get (always) crash also with the game itself when I start it with full validation (no crash with just core validation, but that could be just a fluke).  Without validation game works fine. Binaries & libs are same as yesterday.

Game backtrace looks different than from the replay and I don't see any validation warnings from the game itself (does it redirect them somewhere?):
```
Thread 1 "dota2" received signal SIGSEGV, Segmentation fault.
anv_CreateSwapchainKHR (_device=0x5555565f9400, pCreateInfo=0x7fffffffca40, pAllocator=0x0, pSwapchain=0x5555567e1390)
    at ../../../../src/intel/vulkan/anv_wsi.c:264
264	   struct wsi_interface *iface =
(gdb) list
261	{
262	   ANV_FROM_HANDLE(anv_device, device, _device);
263	   ANV_FROM_HANDLE(_VkIcdSurfaceBase, surface, pCreateInfo->surface);
264	   struct wsi_interface *iface =
265	      device->instance->physicalDevice.wsi_device.wsi[surface->platform];
(gdb) print *device->instance->physicalDevice.wsi_device.wsi
value has been optimized out
(gdb) info locals
iface = <optimized out>
swapchain = 0x1
alloc = <optimized out>
result = <optimized out>
(gdb) print _device
$7 = (VkDevice) 0x5555565f9400
(gdb) print surface->platform
value has been optimized out
(gdb) print pCreateInfo 
$10 = (const VkSwapchainCreateInfoKHR *) 0x7fffffffca40
(gdb) print pCreateInfo->surface
$8 = (VkSurfaceKHR) 0x1
(gdb) disassemble 
Dump of assembler code for function anv_CreateSwapchainKHR:
...
   0x00007fffec02228a <+42>:	mov    0x18(%rsi),%rdi
   0x00007fffec02228e <+46>:	test   %rbx,%rbx
=> 0x00007fffec022291 <+49>:	mov    (%rdi),%ecx
(gdb) info registers rsi rdi ecx
rsi            0x7fffffffca40	140737488341568
rdi            0x1	1
ecx            0x567e1390	1451103120
```

-> pCreateInfo->surface is invalid.

**Comment by eero-t** [comment]
Wrote a script to run DOTA2 directly from console to make sure I see validation messages.

Today's git versions (with Mesa now supporting tessellation also for Vulkan):
* Was able to get trace of DOTA2's gameplay .dem recording replay (one from Phoronix DOTA2 PTS test)
* vkreplay for that trace segfaults if run as-is, or with full validation (bad swapchain index), it doesn't segfault if run with core validation or under strace.  I'm assuming why this changed is just timing related randomness
* During tracing, sometimes frames are not displayed in right order, one often sees very old frames (maybe even tens of frames old?), like earlier

There are a lot of these kind of errors during replay from core validation:
```
DS(ERROR): object: 0x0 type: 0 location: 3276 msgCode: 1426: Invalid Pipeline CreateInfo State: VK_PRIMITIVE_TOPOLOGY_PATCH_LIST primitive topology used with patchControlPoints value 0. patchControlPoints should be >0 and <=32. For more information refer to Vulkan Spec Section '21.7. Tessellation Pipeline State' which states 'patchControlPoints must be greater than zero and less than or equal to VkPhysicalDeviceLimits::maxTessellationPatchSize' (https://www.khronos.org/registry/vulkan/specs/1.0-extensions/xhtml/vkspec.html#VkPipelineTessellationStateCreateInfo)
DS(ERROR): object: 0x77ce950 type: 19 location: 2975 msgCode: 50: At Draw time the active render pass (0x779d880) is incompatible w/ gfx pipeline (0x77ce950) that was created w/ render pass (0x6ce14d0) due to: color attachments at index 0 of subpass index 0 are not compatible.
```
